### PR TITLE
Add cassanknex

### PR DESCRIPTION
**Pull request by bioball**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
   - This is not working for me. However, I can verify that the typings match up to what the module actually exports.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/cassanknex/cassanknex-tests.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import * as cassanknex from "cassanknex";
```
[the remark this answers is not in the document]
Comment (bioball) [reply types/cassanknex/cassanknex-tests.ts]: Done!

**Comment by bioball** [comment]
Hey @DanielRosenwasser can you give a thumbs for this? Thanks!

File: types/cassanknex/tslint.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+	"extends": "dtslint/dt.json",
+	"rules": {
+		"space-before-function-paren": [true, "always"],
```
Comment (DanielRosenwasser) [inline types/cassanknex/tslint.json]: Can you turn off these lint rule exceptions so that this file is just an extension to the main lint rules?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by bioball** [comment]
Hey @DanielRosenwasser 

I've enabled most of the dtslint rules, but I've left "no-unnecessary-generics" and "strict-export-declare-modifiers" turned off.

In this case, "no-unnecessary-generics" lints against wanted behavior of these typings. The goal is to allow uses to write their queries like so, and have it eventually type-checek:

```ts
export interface IUser {
  id: number,
  name: string,
  age: num ber
}
knex()
  .select<IUser>('name')
  .where('id', '=', 5);
```

Without that generic, this typechecking is not possible.

As for "strict-export-declare-modifiers", I'm not really sure how to fix that error :/  this file is based off of TypeScript's own recommendations for how to annotate a [function module](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-function-d-ts.html).

**Comment by bioball** [comment]
@DanielRosenwasser pinging again here

**Comment by bioball** [comment]
@DanielRosenwasser @RyanCavanaugh can I get a thumb on this one?

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by sheetalkamat** [comment]
@DanielRosenwasser I don't think we allow any lint rules in any of the new declaration files. We generally advise users to correct those instead. Is this exception?

File: types/cassanknex/tslint.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+	"extends": "dtslint/dt.json",
+  "rules": {
+    "no-unnecessary-generics": false,
```
Comment (sheetalkamat) [inline types/cassanknex/tslint.json]: Please remove these rules and apply them locally for the line if absolutely needed

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/cassanknex/tslint.json (hunk 3 above)
Comment (bioball) [reply types/cassanknex/tslint.json]: Hey @sheetalkamat, can you explain how I'm supposed to write this typing and follow the `strict-export-declare-modifiers` rule?

**Pull request closed without merging**